### ISSUE TITLE
关于LIF神经元膜电压的值？

### ISSUE THREAD

**Issue by ROY-SNN** [fangwei123456/spikingjelly#220]
您好，最近我在尝试利用classify_dvsg.py中的模型，经过修改来处理Nav-Gesture-sit的数据集，训练后想尝试利用visualizing.plot_2d_heatmap可视化一下fc中最后的LIF层的膜电压积累到释放的过程，但是实验发现，膜电压大部分都为负电压，不知道这是否正常？我看使用说明手册中“1.3.3 时间驱动的仿真方式”中 lif = neuron.LIFNode(tau=100.) 举例的图示中，膜电压范围为0~1.0，而在“1.5. 时间驱动：使用单层全连接 SNN 识别 MNIST”中举例的图示中，膜电压大部分为负值。

 

          def __init__(self, channels: int):
             super().__init__()
  
             self.net_name = "Navgesture_sit_SNN_Net"   
  
             conv = []
             conv.extend(Navgesture_sit_SNN_Net2.conv3x3(2, channels))
             conv.append(nn.MaxPool2d(2, 2))
  
             for i in range(6):
                 conv.extend(Navgesture_sit_SNN_Net2.conv3x3(channels, channels))
                 conv.append(nn.MaxPool2d(2, 2))
             self.conv = nn.Sequential(*conv)
  
             self.fc = nn.Sequential(
                 nn.Flatten(),
                 layer.Dropout(0.5),
                 nn.Linear(channels * 2, channels, bias=False),
                 neuron.LIFNode(tau=2.0, surrogate_function=surrogate.ATan(), detach_reset=True),
                 layer.Dropout(0.5),
                 nn.Linear(channels, 50, bias=False),
                 neuron.LIFNode(tau=2.0, surrogate_function=surrogate.ATan(), detach_reset=True)
              )
  
              self.vote = VotingLayer(8)

         def forward(self, x: torch.Tensor):
              v_list = []
              x = x.permute(1, 0, 2, 3, 4) 

              temp = self.fc(self.conv(x[0]))
              print('t=0')
              print('最后一层LIF的膜电压(释放前)：', self.fc[6].v[0])
              print('最后一层LIF的输出脉冲：', temp[0].tolist())
              v_list.append(self.fc[6].v[0].unsqueeze(0))
              out_spikes = self.vote(temp)

              for t in range(1, x.shape[0]):
                   temp = self.fc(self.conv(x[t]))
                   print(f't={t}')
                   print('最后一层LIF的膜电压(释放前)：', self.fc[6].v[0])
                   print('最后一层LIF的输出脉冲：', temp[0].tolist())
                   v_list.append(self.fc[6].v[0].unsqueeze(0))
                   out_spikes += self.vote(temp)

              v_list = torch.cat(v_list)
              visualizing.plot_2d_heatmap(array=np.asarray(v_list), title='Membrane Potentials',xlabel = 'Simulating Step',
                                        ylabel = 'Neuron Index', int_x_ticks = True, x_max = len(np.asarray(v_list)), dpi = 200)
              return out_spikes / x.shape[0]

`
[https://static01.imgkr.com/temp/2153420c5f9d46779eeb55dd5db9780b.png](url)

**Comment by fangwei123456** [comment]
最后一层正确类别的神经元输出1，其他神经元被强烈抑制，所以负电压是很正常的。如果你找一个分错的样本，应该能看到不太一样的结果。

**Comment by ROY-SNN** [comment]
谢谢您的及时回复，还有一个问题：“functional.reset_net(net)”的作用是手动重置LIF神经元的膜电压值吗？源码中训练的时候是每一批量后执行一次functional.reset_net(net)。我在利用训练好的模型进行预测的时候，每预测结束一个样本就执行一次functional.reset_net(net)和不执行functional.reset_net(net)，两次的膜电压不一样，这对模型的预测影响大吗？哪一种策略比较好？

**Comment by fangwei123456** [comment]
> “functional.reset_net(net)”的作用是手动重置LIF神经元的膜电压值吗？

是的。

> 每预测结束一个样本就执行一次functional.reset_net(net)和不执行functional.reset_net(net)，两次的膜电压不一样，这对模型的预测影响大吗？

一定要在输入下一个样本之前，reset整个网络，因为整个网络的状态是由上一个输入的样本生成的，和下一个样本没有关系。reset将整个网络的状态恢复到初始值。
